tests/test_update_ssh_keys.rs

```diff
@@ -0,0 +1,622 @@
+    // assert_keys: Vec of keys to be asserted after add_keys() succeeded.
+    fn add_one_ssh_key(ssh_dir: &str, pubkeyname: &str, testname: &str, is_force: bool, is_replace: bool, assert_keys: Vec<&str>) -> Result<(), ErrorKind> {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        let res = aks.add_keys(testname, keys.clone(), is_replace, is_force);
+
+        match res {
+            Ok(keys) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
+                Ok(())
+            }
+            Err(Error(ErrorKind::KeysDisabled(name), _)) => {
+                println!("Skipping add {}, disabled.", name);
+                Ok(())
+            }
+            Err(Error(ErrorKind::KeysExist(_), _)) => {
+                println!("Skipping add {}, already exists.", testname);
+                Ok(())
+            }
+            Err(err) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
+                Err(err.into())
+            }
+        }
+    }
+
+    // A wrapper for deleting an ssh key.
+    //
+    // pubkeyname: name of ssh public key, like valid1, bad
+    // testname: key name given as cmdline args, like "one", "two"
+    // assert_keys: Vec of keys to be asserted after add_keys() succeeded.
+    fn del_one_ssh_key(ssh_dir: &str, pubkeyname: &str, testname: &str) {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        for key in aks.remove_keys(testname) {
+            if let AuthorizedKeyEntry::Valid {
+                ref key,
+            } = key
+            {
+                assert!(key.to_fingerprint_string().contains(&*FINGERPRINTS[pubkeyname]));
+
+                // must write before checking if the file exists, because before writing,
+                // authorized_keys.d is not in sync.
+                aks.write().expect("failed to update authorized keys directory");
+
+                let authkeyone: PathBuf = PathBuf::from(format!("{}/authorized_keys.d/{}", ssh_dir, testname));
+
+                assert!(!authkeyone.exists());
+
+                let mut svec: Vec<&str> = Vec::new();
+                svec.push(pubkeyname);
+                assert_has_keys(ssh_dir, svec, true);
```

Comment:
I think part of the problem is that this function does too much compared to the same section in `main.rs`:
```
 118   │         Command::Delete { name } => {
 119   │             println!("Removing {}:", name);
 120   │             for key in aks.remove_keys(&name) {
 121   │                 if let AuthorizedKeyEntry::Valid { ref key } = key {
 122   │                     println!("{}", key.to_fingerprint_string());
 123   │                 }
 124   │             }
 125   │         }
 […]
  149   │     aks.write()
 150   │         .chain_err(|| "failed to update authorized keys directory")?;
 151   │     aks.sync()
 152   │         .chain_err(|| "failed to update authorized keys")?;
```
I think it's better to stick to this and move the assertions out here and into the test functions, and not as part of this helper function.